verify/verify.go

```diff
@@ -322,65 +369,73 @@ func validateCRLlink(x *x509.Certificate, product, role string) error {
 	if cert.Version != 3 {
-		return nil, fmt.Errorf("VCEK certificate version is %v, expected 3", cert.Version)
+		return nil, fmt.Errorf("%v certificate version is %v, expected 3", key, cert.Version)
 	}
 	// Signature algorithm: RSASSA-PSS
 	// Signature hash algorithm sha384
 	if cert.SignatureAlgorithm != x509.SHA384WithRSAPSS {
-		return nil, fmt.Errorf("VCEK certificate signature algorithm is %v, expected SHA-384 with RSASSA-PSS", cert.SignatureAlgorithm)
+		return nil, fmt.Errorf("%v certificate signature algorithm is %v, expected SHA-384 with RSASSA-PSS", key, cert.SignatureAlgorithm)
 	}
 	// Subject Public Key Info ECDSA on curve P-384
 	if cert.PublicKeyAlgorithm != x509.ECDSA {
-		return nil, fmt.Errorf("VCEK certificate public key type is %v, expected ECDSA", cert.PublicKeyAlgorithm)
+		return nil, fmt.Errorf("%v certificate public key type is %v, expected ECDSA", key, cert.PublicKeyAlgorithm)
 	}
 	// Locally bind the public key any type to allow for occurrence typing in the switch statement.
 	switch pub := cert.PublicKey.(type) {
 	case *ecdsa.PublicKey:
 		if pub.Curve.Params().Name != "P-384" {
-			return nil, fmt.Errorf("VCEK certificate public key curve is %s, expected P-384", pub.Curve.Params().Name)
+			return nil, fmt.Errorf("%v certificate public key curve is %s, expected P-384", key, pub.Curve.Params().Name)
 		}
 	default:
-		return nil, fmt.Errorf("VCEK certificate public key not ecdsa PublicKey type %v", pub)
+		return nil, fmt.Errorf("%v certificate public key not ecdsa PublicKey type %v", key, pub)
 	}
 
-	if err := ValidateVcekCertSubject(cert.Subject); err != nil {
+	if err := validateKDSCertSubject(cert.Subject, key); err != nil {
 		return nil, err
 	}
-	exts, err := kds.VcekCertificateExtensions(cert)
+	exts, err := kds.CertificateExtensions(cert, key)
 	if err != nil {
 		return nil, err
 	}
-	if err := validateVcekExtensions(exts); err != nil {
+	if err := validateExtensions(exts, key); err != nil {
 		return nil, err
 	}
 	return exts, nil
 }
 
-func validateVcekCertificateProductSpecifics(r *trust.AMDRootCerts, cert *x509.Certificate, opts *Options) error {
-	if err := ValidateVcekCertIssuer(r, cert.Issuer); err != nil {
+func validateKDSCertificateProductSpecifics(r *trust.AMDRootCerts, cert *x509.Certificate, key abi.ReportSigner, opts *Options) error {
+	if err := validateKDSCertIssuer(r, cert.Issuer, key); err != nil {
 		return err
 	}
-	if _, err := cert.Verify(*r.X509Options(opts.Now)); err != nil {
-		return fmt.Errorf("error verifying VCEK certificate: %v (%v)", err, r.ProductCerts.Ask.IsCA)
+	// ica: Intermediate Certificate Authority.
+	ica := r.ProductCerts.Ask
+	if key == abi.VlekReportSigner {
+		ica = r.ProductCerts.Asvk
+	}
+	if ica == nil {
+		return fmt.Errorf("root of trust missing intermediate certificate authority certificate for key %v", key)
+	}
+	if _, err := cert.Verify(*r.X509Options(opts.Now, key)); err != nil {
```

Comment:
```suggestion
	tmp := r.X509Options(opts.Now, key)
	if tmp == nil {
		return fmt.Errorf("internal error: could not get X509 options for %v", key)
	}
	if _, err := cert.Verify(*tmp); err != nil {
```

Not sure how to call that variable ^^

Reply:
Ah yes, uninitialized intermediate certs will lead to a nil dereference. Good catch.